src/vs/workbench/api/browser/mainThreadDebugService.ts

```diff
@@ -450,21 +458,27 @@ export class MainThreadDebugService implements MainThreadDebugServiceShape, IDeb
 					condition: fbp.condition,
 					hitCondition: fbp.hitCondition,
 					logMessage: fbp.logMessage,
-					functionName: fbp.name
+					functionName: fbp.name,
+					mode: fbp.mode
 				} satisfies IFunctionBreakpointDto;
 			} else if ('src' in bp) {
 				const dbp: IDataBreakpoint = bp;
 				return {
 					type: 'data',
 					id: dbp.getId(),
-					dataId: dbp.src.type === DataBreakpointSetType.Variable ? dbp.src.dataId : dbp.src.address,
+					source: dbp.src.type === DataBreakpointSetType.Variable ? new ResolvedDataBreakpointSource(dbp.src.dataId, dbp.src.canPersist)
+						: dbp.src.type === DataBreakpointSetType.Address ? new AddressDataBreakpointSource(dbp.src.address, dbp.src.bytes)
+							: dbp.src.type === DataBreakpointSetType.Expression ? new ExpressionDataBreakpointSource(dbp.src.expression)
+								: dbp.src.type === DataBreakpointSetType.FrameScoped ? new FrameScopedDataBreakpointSource(dbp.src.frameId, dbp.src.expression)
+									: dbp.src.type === DataBreakpointSetType.VariableScoped ? new VariableScopedDataBreakpointSource(dbp.src.variablesReference, dbp.src.variable)
+										: new ResolvedDataBreakpointSource('-1'), // should not happen
```

Comment:
Here and above, you can use `assertNever` if TS agrees it can never be hit

Reply:
Reworked the whole flow a bit so that now we shouldn't run into that problem anymore.